lib/translator/status.cc

```diff
@@ -0,0 +1,248 @@
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedFailedFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedMissingFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kInvalidNamespaceOrFormat:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc =
+          scsi_defs::AdditionalSenseCode::kAccessDeniedInvalidLuIdentifier;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kAccessDeniedInvalidLuIdentifier;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kLbaOutOfRange:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc = scsi_defs::AdditionalSenseCode::kLbaOutOfRange;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::kLbaOutOfRange;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kNamespaceNotReady:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kNotReady;
+      result.asc = scsi_defs::AdditionalSenseCode::
+          kLogicalUnitNotReadyCauseNotReportable;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kLogicalUnitNotReadyCauseNotReportable;
+      break;
+    default:
+      DebugLog("No SCSI translation for NVMe status with code %#x",
+               static_cast<uint8_t>(sc));
+  }
+  return result;
+}
+
+// Section 7.2
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+ScsiStatus StatusToScsi(nvme_defs::CommandSpecificStatusCode sc) {
+  ScsiStatus result = kDefaultScsiStatus;
+
+  switch (sc) {
+    case nvme_defs::CommandSpecificStatusCode::kCompletionQueueInvalid:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      break;
+    case nvme_defs::CommandSpecificStatusCode::kInvalidQueueIdentifier:
+    case nvme_defs::CommandSpecificStatusCode::kInvalidQueueSize:
+    case nvme_defs::CommandSpecificStatusCode::kAbortCommandLimitExceeded:
+    case nvme_defs::CommandSpecificStatusCode::kAsyncEventRequestLimitExceeded:
+    case nvme_defs::CommandSpecificStatusCode::kInvalidFirmwareSlot:
+    case nvme_defs::CommandSpecificStatusCode::kInvalidFirmwareImage:
+    case nvme_defs::CommandSpecificStatusCode::kInvalidInterruptVector:
+    case nvme_defs::CommandSpecificStatusCode::kInvalidLogPage:
+      // Translation unspecified
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      break;
```

Comment:
Here, the code for `kCompletionQueueInvalid` is the same as for the next clumped cases. This seems deliberate (I'm guessing to differentiate the explicit conversion from undefined ones?). I would extend your `// Translation unspecified` comment (or add a new comment above) to call out why `kCompletionQueueInvalid` gets its own section, even when the actual logic is equivalent.